### PR TITLE
Command execution events + subrequests support + volume_id setting

### PR DESCRIPTION
**Pull request by nicolasmure**
Hi,

This PR provides events on elFinder command execution (pre and post).
It also allows to make subrequests to relaunch an elFinder command.
A new setting has been added : `volume_id`, necessary to keep the volume identifier when making subrequests on the same volume than the main request.

Related to [# 19](https://github.com/helios-ag/ElFinderPHP/pull/19) on the lib side.

### CODE REVIEW

**Comment by nicolasmure** [comment]
I've made this PR as I talk about in #126 .
This is how I handle multiple thumbnails with new events and LiipImagineBundle : 

```php
namespace AppBundle\EventListener\ElFinder;

use FM\ElfinderBundle\Event\ElFinderEvents;

abstract class ExecutionListener
{
    /**
     * Performs a subrequest to elFinderBundle to return informations.
     * @param  ElFinderPostExecutionEvent $event
     * @return array
     */
    protected function getInfos(ElFinderEvents $event)
    {
        $queryParameters = $event->getRequest()->query->all();
        $queryParameters['cmd'] = 'info';

        if (isset($queryParameters['target']) && !isset($queryParameters['targets'])) {
            $queryParameters['targets'] = array($queryParameters['target']);
            unset($queryParameters['target']);
        }

        $jsonResponse = $event->subRequest(array(
            'instance' => $event->getInstance(),
            'homeFolder' => $event->getHomeFolder()
        ), $queryParameters);

        $filesInfos = json_decode($jsonResponse->getContent());
        return $filesInfos->files;
    }
}
```

```php
<?php

namespace AppBundle\EventListener\ElFinder;

use Liip\ImagineBundle\Controller\ImagineController;
use FM\ElfinderBundle\Event\ElFinderPostExecutionEvent;
use Symfony\Component\HttpFoundation\Request;

class PostExecutionListener extends ExecutionListener
{
    /**
     * LiipImagineBundle Controller
     * @var ImagineController
     */
    private $imagineController;


    /**
     * Constructor.
     * @param ImagineController $imagineController
     */
    public function __construct(ImagineController $imagineController)
    {
        $this->imagineController = $imagineController;
    }

    /**
     * Manages thumbnails creation.
     * @param  ElFinderPostExecutionEvent $event
     */
    public function onPostExecute(ElFinderPostExecutionEvent $event)
    {
        if (!$event->hasErrors()) {
            if ($event->getCommand() == 'tmb') {
                $this->createThumbnails($event);
            }
        }
    }

    /**
     * Creates picture's thumbnails with LiipImagineBundle.
     * @param ElFinderPostExecutionEvent $event
     */
    private function createThumbnails(ElFinderPostExecutionEvent $event)
    {
        $imagesInfos = $this->getInfos($event);
        $request = new Request();

        foreach ($imagesInfos as $imageInfo) {
            $path = '/'.$imageInfo->path;

            $this->imagineController->filterAction($request, $path, 'tmb_64');
            $this->imagineController->filterAction($request, $path, 'tmb_128');
        }
    }
}
```

```php
namespace AppBundle\EventListener\ElFinder;

use Liip\ImagineBundle\Imagine\Cache\CacheManager;
use FM\ElfinderBundle\Event\ElFinderPreExecutionEvent;

class PreExecutionListener extends ExecutionListener
{
    /**
     * LiipImagineBundle CacheManager
     * @var CacheManager
     */
    private $cacheManager;


    /**
     * Constructor.
     * @param CacheManager $cacheManager
     */
    public function __construct(CacheManager $cacheManager)
    {
        $this->cacheManager = $cacheManager;
    }

    /**
     * Manages thumbnails delete.
     * @param  ElFinderPreExecutionEvent $event
     */
    public function onPreExecute(ElFinderPreExecutionEvent $event)
    {
        $command = $event->getCommand();
        if ($command == 'rm' || $command == 'rename' || $command == 'resize') {
            $this->removeThumbnails($event);
        }
    }

    /**
     * Removes picture's thumbnails created with LiipImagineBundle.
     * @param ElFinderPreExecutionEvent $event
     */
    private function removeThumbnails(ElFinderPreExecutionEvent $event)
    {
        $imagesInfos = $this->getInfos($event);

        foreach ($imagesInfos as $imageInfo) {
            $path = '/'.$imageInfo->path;

            $this->cacheManager->remove($path);
        }
    }
}
```

File: Event/ElFinderPreExecutionEvent.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     * @return Symfony\Component\HttpFoundation\Response A Response instance
+     */
+    public function subRequest(array $path, array $query)
+    {
+        $path['_controller'] = 'FMElfinderBundle:ElFinder:load';
+        $subRequest = $this->request->duplicate($query, null, $path);
+        
+        return $this->httpKernel->handle($subRequest, HttpKernelInterface::SUB_REQUEST);
+    }
+
```
Comment (helios-ag) [inline Event/ElFinderPreExecutionEvent.php]: So this became incompatible with 2.3?

Comment (nicolasmure) [inline Event/ElFinderPreExecutionEvent.php]: Yes, compatible with ElFinderPHP pull request [# 19](https://github.com/helios-ag/ElFinderPHP/pull/19)

**Comment by nicolasmure** [comment]
Hi @helios-ag ,

Do you have some news about it?
I think this PR could be used in favor of #126

**Comment by helios-ag** [comment]
Hi @nicolasmure, well it looks better of course. Anyway i support 2.3 in a separate branch. I could accept this if you can rebase PR. I will hold accepting PRs if you wish.

**Comment by nicolasmure** [comment]
OK, I was just wondering if you were interested by this or not.
I've rebased on master, but I can still work from my fork using composer's vcs if you don't want to merge it now.
By the way, I think PR #126 could be closed.

Thank you :wink:

**Comment by nicolasmure** [comment]
thank you @helios-ag :wink: 

please do not forget that this PR is related to [PR # 19](https://github.com/helios-ag/ElFinderPHP/pull/19 "returning path when using 'info' command + provided GET query parameters") on the lib side

**Comment by helios-ag** [comment]
:+1:

**Pull request merged**